crates/interpreter/src/state.rs

```diff
@@ -0,0 +1,246 @@
+use std::ops::{Add, BitAnd, BitOr, BitXor, Mul, Neg, Not, Sub};
+
+use sonatina_ir::{
+    insn::{BinaryOp, CastOp, UnaryOp},
+    module::FuncRef,
+    Block, DataLocationKind, Immediate, InsnData, Module,
+};
+
+use crate::{Frame, Literal, ProgramCounter};
+
+struct State {
+    module: Module,
+    frames: Vec<Frame>,
+    pc: ProgramCounter,
+    prev_block: Option<Block>,
+}
+
+impl State {
+    // the cpu
+    pub fn new(module: Module, entry_func: FuncRef) -> Self {
+        let func = &module.funcs[entry_func];
+        let pc = ProgramCounter::new(entry_func, &func.layout);
+        let entry_frame = Frame::new(func, pc, vec![]);
+        let frames = vec![entry_frame];
+
+        Self {
+            module,
+            frames,
+            pc,
+            prev_block: None,
+        }
+    }
+
+    pub fn repl(mut self) -> Option<Literal> {
+        loop {
+            if let Some(arg) = self.step() {
+                return arg;
+            }
+        }
+    }
+
+    pub fn step(&mut self) -> Option<Option<Literal>> {
```

Comment:
It'd be nicer to define an enum to represent each step result since this is a public method, especially since this method is expected to be heavily used by users(repl, verification, etc.).